### PR TITLE
SameSite cookie at application level

### LINKED ISSUE #10925: SameSite cookie at application level
Add code to WebContainer for design in #10788

### PR DESCRIPTION
**Pull request by pmd1nh**
#10788

fixes #10925

### CODE REVIEW

**Comment by pnicolucci** [comment]
#build , just curious what the state of the build is with this PR in its current state.

**Comment by pnicolucci** [comment]
#run-libby-bot

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 1]
```diff
@@ -1603,3 +1603,16 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
...
+duplicate.samesite.attribute.for.cookie.name.useraction=The deployment descriptor's context parameters cookie names and values for 'SameSiteCookies_Lax', 'SameSiteCookies_None', and 'SameSiteCookies_Strict' must be unique across all three lists. 
+
+duplicate.reported.for.samesite.attribute.cookie.name=CWWWC0004W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] was found previously.  No SameSite attribute will be added for this cookie.
+duplicate.reported.for.samesite.attribute.cookie.name.explanation=A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
+duplicate.reported.for.samesite.attribute.cookie.name.useraction=The deployment descriptor's context parameters cookie names and values for 'SameSiteCookies_Lax', 'SameSiteCookies_None', and 'SameSiteCookies_Strict' must be unique across all three lists. 
+
+too.many.wildcard.configured.for.samesite.cookie.name=CWWWC0005W: Too many wildcard cookie name configured for SameSite attribute.  No attribute will be added for any cookie.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: **change**
Too many wildcard cookie name configured for SameSite attribute.  No attribute will be added for any cookie.
**to**
Too many wildcard cookie names are configured for multiple SameSite attributes.  No attributes are added for any cookie.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 2]
```diff
@@ -1603,3 +1603,16 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
...
+too.many.wildcard.configured.for.samesite.cookie.name.explanation=Too many wildcard cookie name configured for multiple SameSite attribute.  There should be maxinum one wildcard cookie name configured for SameSite attribute.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: **change**
Too many wildcard cookie name configured for multiple SameSite attribute.  There should be maxinum one wildcard cookie name configured for SameSite attribute.
**to**
Too many wildcard cookie names are configured for multiple SameSite attributes. You cannot configure more than one wildcard cookie name for each SameSite attribute.

Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: **change**
Review the deployment decriptor and limit to one wildcard cookie name for SameSite attribute.
**to**
Review the deployment descriptor configuration and make sure each SameSite attribute has only one wildcard cookie name.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 3]
```diff
@@ -1603,3 +1603,16 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
 no.servlet.defined.for.servlet.filter.name.mapping=CWWWC0002W: No servlet definition is found for the {0} servlet name in the {1} filter mapping.
 no.servlet.defined.for.servlet.filter.name.mapping.explanation=The servlet name is used in filter mapping, but its definition has not been defined.
 no.servlet.defined.for.servlet.filter.name.mapping.useraction=Define the servlet which is used in the servlet name filter mapping.
+
+# {0} Cookie name that has conflict SameSite attribute
+duplicate.samesite.attribute.for.cookie.name=CWWWC0003W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] is configured with new value [{2}].  No SameSite attribute will be added for this cookie.
```
[the remark this answers is not in the document]
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: CWWWC003W - to inform the user that they can not add a "different" SameSite attribute value to an existing cookie name.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 4]
```diff
@@ -1603,3 +1603,16 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
...
+duplicate.samesite.attribute.for.cookie.name.explanation=A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
+duplicate.samesite.attribute.for.cookie.name.useraction=The deployment descriptor's context parameters cookie names and values for 'SameSiteCookies_Lax', 'SameSiteCookies_None', and 'SameSiteCookies_Strict' must be unique across all three lists. 
+
+duplicate.reported.for.samesite.attribute.cookie.name=CWWWC0004W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] was found previously.  No SameSite attribute will be added for this cookie.
```
[the remark this answers is not in the document]
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: Problem is same cookie name but have multiple values.  We removed that cookie name and saved it to a duplicate cookie list (that first detection get reported in the above CWWWC003W).  

When a cookie is parsed and we see that it already marked duplicate (from the duplicate list), CWWWC0004W is used to tell user that this cookie has reported as duplicate and the SameSite attribute will not be added for that cookie.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 2 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: I changed it to:

too.many.wildcard.configured.for.samesite.cookie.name.explanation=Too many wildcard cookie names are configured for multiple SameSite attributes. You can configure only one wildcard cookie name across all SameSite attribute lists.

too.many.wildcard.configured.for.samesite.cookie.name.useraction=Review the deployment descriptor configuration and make sure only one wildcard cookie name is used across all SameSite attribute lists.

==============
(We allow only one wildcard cookie name to be configured across all three list of SameSite attributes)

**Comment by pmd1nh** [comment]
@dmuelle I updated and pushed up the modified messages.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: ok- in that case
**change**
A duplicate cookie name [{0}] with SameSite attribute value [{1}] is configured with new value [{2}].  No SameSite attribute will be added for this cookie.

**to**
The {0} cookie name is already configured with the {1} SameSite attribute value.  The  {2} new SameSite attribute value is not added for this cookie name.

unless adding another SS value causes both attribute values to be removed...

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: ok
**change**
A duplicate cookie name [{0}] with SameSite attribute value [{1}] was found previously.  No SameSite attribute will be added for this cookie.
**to**
The {0} cookie name is a duplicate.  No SameSite attribute is added for this cookie.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 5]
```diff
@@ -1603,3 +1603,18 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
 no.servlet.defined.for.servlet.filter.name.mapping=CWWWC0002W: No servlet definition is found for the {0} servlet name in the {1} filter mapping.
 no.servlet.defined.for.servlet.filter.name.mapping.explanation=The servlet name is used in filter mapping, but its definition has not been defined.
 no.servlet.defined.for.servlet.filter.name.mapping.useraction=Define the servlet which is used in the servlet name filter mapping.
+
+# {0} Cookie name that has conflict SameSite attribute
+# {1} current SameSite attribute value
+# {2} new SameSite attribute value
+duplicate.samesite.attribute.for.cookie.name=CWWWC0003W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] is configured with new value [{2}].  No SameSite attribute will be added for this cookie.
+duplicate.samesite.attribute.for.cookie.name.explanation=A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: **change**
A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
**to**
A cookie name cannot have multiple SameSite attribute values.  No attribute is added for the cookie.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 6]
```diff
@@ -1603,3 +1603,18 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
...
+duplicate.samesite.attribute.for.cookie.name.useraction=The deployment descriptor's context parameters cookie names and values for 'SameSiteCookies_Lax', 'SameSiteCookies_None', and 'SameSiteCookies_Strict' must be unique across all three lists.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: Are  
`SameSiteCookies_Lax`, `SameSiteCookies_None`, and `SameSiteCookies_Strict` 
SameSite attributes?

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 7]
```diff
@@ -1603,3 +1603,18 @@ exception.occurred.while.running.ServletContainerInitializers.onStartup.useracti
...
+
+# {0} Cookie name that has conflict SameSite attribute
+# {1} current SameSite attribute value
+# {2} new SameSite attribute value
+duplicate.samesite.attribute.for.cookie.name=CWWWC0003W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] is configured with new value [{2}].  No SameSite attribute will be added for this cookie.
+duplicate.samesite.attribute.for.cookie.name.explanation=A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
+duplicate.samesite.attribute.for.cookie.name.useraction=The deployment descriptor's context parameters cookie names and values for 'SameSiteCookies_Lax', 'SameSiteCookies_None', and 'SameSiteCookies_Strict' must be unique across all three lists. 
+
+duplicate.reported.for.samesite.attribute.cookie.name=CWWWC0004W: A duplicate cookie name [{0}] with SameSite attribute value [{1}] was found previously.  No SameSite attribute will be added for this cookie.
+duplicate.reported.for.samesite.attribute.cookie.name.explanation=A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: **change**
A cookie name must not have multiple SameSite attribute values.  The attribute will not be added for the cookie.

**to**
A cookie name cannot have multiple SameSite attribute values.  The attribute is not added for the cookie.

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 2 above)
Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: that works

**Comment by pnicolucci** [comment]
@pmd1nh  what more is remaining here? Do we just need to rebase and get a good build? I'd like to get this delivered sooner rather than later so I can get some tests executing. A couple of comments:

**Comment by pmd1nh** [comment]
#build  #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.1,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#run-libby-bot

**Pull request merged**